tests/spki_tests.rs

```diff
@@ -0,0 +1,701 @@
+extern crate ring;
+extern crate untrusted;
+extern crate rustc_serialize;
+
+use std::io::BufRead;
+use std::path::Path;
+use std::vec::Vec;
+use std::io::Read;
+use std::fs::File;
+use rustc_serialize::base64::FromBase64;
+use ring::der;
+use ring::signature::spki;
+use ring::signature::spki::VerifyWithSPKIError;
+#[cfg(feature = "rsa_signing")]
+use ring::signature;
+#[cfg(feature = "rsa_signing")]
+use ring::rand;
+
+// TODO: The expected results need to be modified for SHA-1 deprecation.
+
+macro_rules! test_verify_signature_pem {
+    ($fn_name:ident, $file_name:expr, $signature_alg:expr, $expected_result:expr) => {
+        #[test]
+        fn $fn_name() {
+            test_verify_signature_pem($file_name, $signature_alg, $expected_result);
+        }
+    }
+}
+
+fn test_verify_signature_pem(file_name: &str,
+                             signature_algorithm: &'static spki::Algorithm,
+                             expected_result: Result<(), VerifyWithSPKIError>) {
+    let tsd = parse_test_signed_data(file_name);
+    let spki_value = untrusted::Input::from(&tsd.spki);
+
+    let signature = untrusted::Input::from(&tsd.signature);
+    let signature = signature.read_all(VerifyWithSPKIError::BadDER, |input| {
+        der::bit_string_with_no_unused_bits(input)
+            .map_err(|_| VerifyWithSPKIError::BadDER)
+    }).unwrap();
+
+    assert_eq!(expected_result,
+    spki::verify(signature_algorithm,
+                 spki_value,
+                 untrusted::Input::from(&tsd.data),
+                 signature));
+}
+
+macro_rules! test_rsa_sign_verify_spki {
+    ($fn_name:ident, $signing_encoding:expr, $algorithm:expr, $key_id:expr) => {
+        #[cfg(feature = "rsa_signing")]
+        #[test]
+        fn $fn_name() {
+            rsa_sign_and_verify_spki($signing_encoding, $algorithm, $key_id);
+        }
+    }
+}
+
+macro_rules! test_rsa_verify_sig_file_spki {
```

Comment:
That's straightforward for the case where the signature should succeed, but less so for the other failure cases covered by the macro (using the other input file, a different digest algorithm, and the other key) since there isn't an easy way to find an equivalent `Algorithm` with, say, a different digest. How would you suggest handling those tests?